### PR TITLE
Add URI#resolve and URI#relativize

### PR DESCRIPTION
**Pull request by straight-shoota**
This PR adds two complementary methods to `URI`:

* `URI#resolve(uri)` resolves `uri` against `self` according to [RFC 3986 section 5.2](https://tools.ietf.org/html/rfc3986#section-5.2.2).
* `URI#relativize(uri)` relativizes `uri` against `self` using `query` and `fragment` components from `uri` and a path component that describes a minimum-difference relative path between the paths of `self` and `uri`.

```cr
URI.parse("http://foo.com/bar/baz").resolve("../quux")                # => http://foo.com/quux
URI.parse("http://foo.com/bar/baz").relativize("http://foo.com/quux") # => ../quux`
```

#7697 contains a similar implementation for `URI#resolve` called `URI#merge`

### CODE REVIEW

**Comment by ysbaddaden** [comment]
I prefer merge. It's more explicit than resolve which is confusing —what is is supposed to resolve?

It's also explicitly named "merge" and "remove dot segments" in RFC 3986 sections 5.2.3 and 5.2.4. The term "resolve" is never used, even in section 5.2.2.

I had to dig the specs to start to understand what relativise could mean. I don't see any use case. I don't think it's frequent enough to be in stdlib.

**Comment by straight-shoota** [comment]
The entire section 5 of  RFC 3986 is titled *URI resolution* and that's the common term for this, also used in many APIs. The RFC only refers to merging of paths.
Calling the entire resolution process `merge` seems to be a Rubyism. I don't think this is a good name, because it's pretty unspecific. Using the more expressive name `resolve` is better.

> I had to dig the specs to start to understand what relativise could mean.

The docs explicitly state it's the inverse of `#resolve`.
`#relativize` is definitely not as commonly used as `#resolve`, so yeah, it might not fit into stdlib. But it's still useful and complements `#resolve`. The implementation is already not very complex and  I suppose the implementation of `#relativize_path` can eventually move to `Path` because having a similar method on `Path` would be nice.

**Comment by oprypin** [comment]
@ysbaddaden is the only one suggesting the name `merge` (and really insisting on it, for whatever reason).

But I think it just doesn't fit. Programmers tend to think of merging as taking the "best" parts of two sequences, possibly by indiscriminately iterating over them.

And that's very misleading in this context because it's explicitly *not* this type of operation.

`resolve` makes a lot of sense to me, and @straight-shoota provides good justification as well.

**Comment by oprypin** [comment]
But I think there's a mistake in the original post. The last example seems to be demonstrating `relativize` but actually calls `resolve`.

**Comment by straight-shoota** [comment]
Fixed

**Comment by bcardiff** [comment]
I think there is still an issue in the opening comment:

It should probably be:

```crystal
URI.parse("http://foo.com/bar/baz").relativize("http://foo.com/quux") # => ../quux`
```

File: src/uri.cr [hunk 1]
```diff
@@ -19,6 +19,28 @@ require "./uri/uri_parser"
 # uri.query  # => "id=30&limit=5"
 # uri.to_s   # => "http://foo.com/posts?id=30&limit=5#time=1305298413"
 # ```
+#
+# ## Resolution and Relativization
+#
+# *Resulution* is the process of resolving one URI against another, *base* URI.
```
Comment (r00ster91) [inline src/uri.cr]:
```suggestion
# *Resolution* is the process of resolving one URI against another, *base* URI.
```

**Comment by straight-shoota** [comment]
Rebased on master

**Comment by straight-shoota** [comment]
Rebased onto master.

File: src/uri.cr (hunk 1 above)
Comment (RX14) [reply src/uri.cr]: This change was lost

Comment (straight-shoota) [reply src/uri.cr]: Thanks

**Pull request merged**